lib/graphql/validation/rules/fields_on_correct_type.ex

```diff
@@ -0,0 +1,96 @@
+
+defmodule GraphQL.Validation.Rules.FieldsOnCorrectType do
+
+  alias GraphQL.Lang.AST.{Visitor, TypeInfo}
+  alias GraphQL.Type.{CompositeType, ObjectType, AbstractType}
+  alias GraphQL.Type
+
+  defstruct name: "FieldsOnCorrectType"
+
+  defimpl Visitor do
+
+    @max_type_suggestions 5
+
+    def enter(_visitor, node, accumulator) do
+      if node.kind == :Field do
+        schema = accumulator[:type_info].schema
+        parent_type = TypeInfo.parent_type(accumulator[:type_info])
+        if parent_type do
+          field_def = TypeInfo.find_field_def(schema, parent_type, node)
+          if !field_def do
+            accumulator = put_in(
+              accumulator[:validation_errors],
+              [  undefined_field_message(schema, node.name.value, parent_type)]
+              ++ accumulator[:validation_errors]
+            )
+          end
+        end
+      end
+      {:continue, accumulator}
+    end
+
+    def leave(_visitor, _node, accumulator) do
+      {:continue, accumulator}
+    end
+
+    defp sibling_interfaces_including_field(schema, type, field_name) do
+      AbstractType.possible_types(type, schema)
+        |> Enum.filter(&is_a_graphql_object_type/1)
```

Comment:
Think about using `Stream` for this pipeline as it should be a drop in replacement http://elixir-lang.org/docs/stable/elixir/Stream.html